### ISSUE TITLE
Code Part 2 - Gradient Descent

### ISSUE THREAD

**Issue by HuwWDay** [Bristol-Training/intro_to_AI#6]
Please don't start work on this until you've met with Huw!

Aim to finish these at latest Monday 28th April. Schedule a meeting halfway through April to touch base. Put all of the notebooks in the Code Part 2 folder in the github.

We need to integrate the theory of SGD and backprop into our code. Building off the [code from part 1](https://github.com/Bristol-Training/intro_to_AI/blob/main/CodePart1/intro_to_AI_zhengzhe.ipynb), basically taking step 5 and 6 more formally and making the code run (they will be removed from code part 1).

- [x] Have a big old MLP class which has the feedforward stuff from part 1 and SGD from part 2 so that the code runs it, have something about tweaking the learning rate and deciding how many epochs to run your code for
- [x] In a new notebook, put some stuff about interpreting these curves, only a little: https://cs231n.github.io/neural-networks-3/#sanitycheck (the graphs in this are great)

![Image](https://github.com/user-attachments/assets/3efed849-5fd1-4472-bb65-7300f0afe579)

- [x] It would be good if you can briefly allude to the differences between doing this process locally and on HPC, but limited details please (some of the feedback is that our course might be a bit too long!)
